common/dpcpp/slab_hash.hpp

```diff
@@ -23,49 +24,118 @@ template <size_t A, size_t B, size_t P> struct DefaultHasher {
+  sycl::device_ptr<SlabNode<T>> next = nullptr;
 };
 
 template <typename T> struct SlabList {
   SlabList() = default;
 
-  sycl::global_ptr<SlabNode<T>> root;
+  sycl::device_ptr<SlabNode<T>> root;
+};
+
+template <typename T> struct HeapMaster {
+  HeapMaster(sycl::queue &q) : _q(q) {
+    _heap = sycl::malloc_device<SlabNode<T>>(CLUSTER_SIZE, q);
+    _head = _heap;
+    sycl::device_ptr<uint32_t> tmp_lock = sycl::malloc_device<uint32_t>(1, q);
+
+    q.single_task([=]() { *tmp_lock = 0; });
+
+    _lock = tmp_lock;
+  }
+
+  ~HeapMaster() { sycl::free(_heap, _q); }
+
+  sycl::device_ptr<SlabNode<T>> malloc_node() {
+    sycl::device_ptr<SlabNode<T>> ret;
+    while (sycl::atomic<uint32_t,
+                        sycl::access::address_space::global_device_space>(_lock)
+               .fetch_or(1)) {
+    }
+    ret = _head;
+    _head++;
+    sycl::atomic<uint32_t, sycl::access::address_space::global_device_space>(
+        _lock)
+        .fetch_and(0);
+    return ret;
+  }
+
+  sycl::device_ptr<uint32_t> _lock = nullptr;
+  sycl::device_ptr<SlabNode<T>> _heap;
+  sycl::device_ptr<SlabNode<T>> _head;
+  sycl::queue &_q;
+};
+template <typename T> struct AllocAdapter {
+  AllocAdapter(size_t bucket_size, T empty, sycl::queue &q, int work_size) : _q(q), _heap(q) {
+    sycl::device_ptr<SlabList<T>> _data_tmp = sycl::malloc_device<SlabList<T>>(bucket_size, q);
+    sycl::device_ptr<uint32_t> _lock_tmp = sycl::malloc_device<uint32_t>(ceil((float)bucket_size / sizeof(uint32_t)), q);
+    _its = sycl::malloc_device<sycl::device_ptr<SlabHash::SlabNode<std::pair<uint32_t, uint32_t>>>>(work_size, q);
+
+    q.parallel_for(bucket_size, [=](auto &i) {
+      *(_data_tmp + i) = SlabList<T>();
+      *(_lock_tmp + i) = 0;
+    });
+
+    _data = _data_tmp;
+    _lock = _lock_tmp;
+  }
+
+  ~AllocAdapter() {
+    sycl::free(_data, _q);
+    sycl::free(_lock, _q);
+  }
+
+  sycl::device_ptr<SlabList<T>> _data;
+  sycl::device_ptr<uint32_t> _lock;
+  HeapMaster<T> _heap;
+  sycl::device_ptr<sycl::device_ptr<SlabHash::SlabNode<std::pair<uint32_t, uint32_t>>>>
+          _its;
+  sycl::queue &_q;
 };
 
 template <typename K, typename T, typename Hash> class SlabHashTable {
 public:
   SlabHashTable() = default;
-  SlabHashTable(K empty, Hash hasher,
-                sycl::global_ptr<SlabList<std::pair<K, T>>> lists,
+  SlabHashTable(K empty,
                 sycl::nd_item<1> &it,
-                sycl::global_ptr<SlabNode<std::pair<K, T>>> &iter)
-      : _lists(lists), _gr(it.get_group()), _it(it), _empty(empty),
-        _hasher(hasher), _iter(iter), _ind(_it.get_local_id()){};
+                SlabHash::AllocAdapter<std::pair<K, T>> &adap)
+      : _lists(adap._data), _gr(it.get_sub_group()), _it(it), _empty(empty),
+         _iter(adap._its[it.get_group().get_id()]), _ind(_it.get_local_id()), _lock(adap._lock),
+        _heap(adap._heap){};
 
   void insert(K key, T val) {
     _key = key;
     _val = val;
```

Comment:
Why need these?

Reply:
I wanted to avoid a large number of passing `key` and `val` as arguments in this way